linera-chain/src/data_types.rs

```diff
@@ -70,13 +70,17 @@ impl Block {
         locations
     }
 
-    /// Returns all the blob IDs referred to in this block's operations.
-    pub fn published_blob_ids(&self) -> HashSet<BlobId> {
+    /// Returns all the published blob IDs in this block's operations.
+    pub fn blob_ids(&self) -> HashSet<BlobId> {
         let mut blob_ids = HashSet::new();
         for operation in &self.operations {
             if let Operation::System(SystemOperation::PublishBlob { blob_id }) = operation {
                 blob_ids.insert(blob_id.to_owned());
             }
+            #[cfg(with_testing)]
+            if let Operation::System(SystemOperation::ReadBlob { blob_id }) = operation {
+                blob_ids.insert(blob_id.to_owned());
+            }
```

Comment:
I think `ReadBlob` is supposed to only simulate a contract runtime `read_blob` call? Then it shouldn't be taken into account here, since from the `Block` alone (without the execution outcome) it can't be known.

Reply:
Good point, will change

Reply:
As discussed, let's not try to check the reads in advance when handling a block _proposal_: It should fail during execution if the blob is missing.

For handling _certificates_, we should use the oracle responses.